2022/Q2/artifacts/pos-model/PoS-model-cubic.md

```diff
@@ -0,0 +1,517 @@
+  var slash = Slash{epoch: evidence.epoch, rate: 0, vpower_fraction: compute_vpower_fraction(evidence.type, vpower)}
+  //enqueue slash (Step 1.1 of cubic slashing)
+  append(enqueued_slashes[evidence.epoch + unbonding_length], slash)
+  //jail validator (Step 1.2 of cubic slashing)
+  validators[validator_address].state[cur_epoch + 1] = jailed
+  remove_from_validator_sets(validator_address, cur_epoch + 1)
+  //freeze validator to prevent delegrators from altering their delegations (Step 1.3 of cubic slashing)
+  validators[validator_address].frozen = true
+}
+```
+
+```go
+func update_voting_power(validator_address, epoch)
+{
+  //compute bonds from total_deltas 
+  var bonds = read_epoched_field(validators[validator_address].total_deltas, epoch, 0)
+  //compute the new voting power
+  var power_after = votes_per_token*bonds
+  //update voting power and return it
+  validators[validator_address].voting_power[epoch] = power_after
+  return power_after
+}
+```
+```go
+func update_total_voting_power(epoch)
+{
+  var total = 0
+  forall (validator in validator_sets[epoch].active U validator_sets[epoch].inactive) do
+    total += validator.voting_power
+  total_voting_power[epoch] = total
+}
+```
+```go
+func update_validator_sets(validator_address, epoch, power_before, power_after)
+{
+  var min_active = first(validator_sets[epoch].active)
+  var max_inactive = last(validator_sets[epoch].inactive)
+  if (power_before >= max_inactive.voting_power) then
+    //if active but it loses power below the max_inactive then move validator to inactive
+    //and promote max_inactive
+    if (power_after < max_inactive.voting_power) then
+      remove(validator_sets[epoch].active, validator_address)
+      add(validator_sets[epoch].active, max_inactive)
+      remove(validator_sets[epoch].inactive, max_inactive.validator)
+      add(validator_sets[epoch].inactive, <validator_address, power_after>)
+    //if active and its change in power does not degrade it to inactive, then update its voting power
+    else
+      remove(validator_sets[epoch].active, validator_address)
+      add(validator_sets[epoch].active, <validator_address, power_after>)
+  else
+    //if inactive and gains power above the min_active, then insert into active and
+    //degrade min_active 
+    if (power_after > min_active.voting_power) then
+      remove(validator_sets[epoch].inactive, validator_address)
+      add(validator_sets[epoch].inactive, min_active)
+      remove(validator_sets[epoch].active, min_active.validator)
+      add(validator_sets[epoch].active, <validator_address, power_after>)
+    //if inactive and its change in power does not promote it to active, then update its voting power
+    else
+      remove(validator_sets[epoch].inactive, validator_address)
+      add(validator_sets[epoch].inactive, <validator_address, power_after>)
+}
+```
+
+```go
+func remove_from_validator_sets(validator_address)
```

Comment:
Two comments (related to each other):
* Isn't this the job of `update_validator_sets`?
* As we are not _marking_ the removed validator, won't the next execution of `update_validator_sets` override the result of this function?
N.B.: It is not totally true that we are not marking the removed validator, because the caller of this function is jailing it, but `update_validator_sets` doesn't take it into account.
So, IMO, we should remove `remove_from_validator_sets`, and modify `update_validator_sets` to remove any jailed validator from the active set.

Reply:
Follow up...
I see you remove the validator from _both_ sets, so I realize `update_validator_sets` won't really override this effects of this.
However, how (and when) will we reintroduce that validator back into one of the two sets?

Reply:
> However, how (and when) will we reintroduce that validator back into one of the two sets?

There is no way to unjail validators at the moment.

Reply:
>So, IMO, we should remove remove_from_validator_sets, and modify update_validator_sets to remove any jailed validator from the active set.

It is cosmetics. Let's first make sure it behaves as expected. We can later merge both functions if we think it is useful.

Reply:
OK, please add a TODO in the model so we don't forget to specify unjailing (after discussing with Heliax, and understanding how they want to model unjailing)

Reply:
Done